### ISSUE TITLE
Crash on moving files when changing base dir

### ISSUE THREAD

**Issue by fm-sys** [cgeo/cgeo#9936]
**Describe the bug:**
I tried to move the base dir via `settings > system` but after granting access to the folder, no progress dialog appeared. The granting access window was still visible, and I saw how the dirs where created. Then the screen turns black and an ANR popped up. Then c:geo crashed and the main activity was visible again...

**Version of c:geo used:**
- 2021.02.01-58cd69b developer build
- Android 10 (emulator)

**Log:**
```
I/ToolbarWidgetWrapper: Progress display unsupported
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/LeakCanary: Watching instance of androidx.fragment.app.FragmentManagerViewModel (androidx.fragment.app.FragmentManagerViewModel received ViewModel#onCleared() callback) with key ada4e66b-2b77-40fb-8c4e-b01bcdba0053
D/LeakCanary: Watching instance of leakcanary.internal.ViewModelClearedWatcher (leakcanary.internal.ViewModelClearedWatcher received ViewModel#onCleared() callback) with key 21fec459-bfb4-4845-96ff-f190559f1ea1
D/LeakCanary: Watching instance of androidx.lifecycle.ReportFragment (androidx.lifecycle.ReportFragment received Fragment#onDestroy() callback) with key 17d376a8-8e08-4a03-826a-c7c35f837dd4
D/LeakCanary: Watching instance of cgeo.geocaching.files.SimpleDirChooser (cgeo.geocaching.files.SimpleDirChooser received Activity#onDestroy() callback) with key 0f31356c-0fa7-4073-a204-0fcbd18f9568
D/LeakCanary: Watching instance of com.android.internal.policy.DecorView (com.android.internal.policy.DecorView received View#onDetachedFromWindow() callback) with key 5f819b70-d5fa-4ade-a909-fb9544296be3
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/LeakCanary: Watching instance of cgeo.geocaching.settings.SettingsActivity (cgeo.geocaching.settings.SettingsActivity received Activity#onDestroy() callback) with key a0feae53-9fce-4430-a69a-364e44017d6b
D/LeakCanary: Watching instance of com.android.internal.policy.DecorView (com.android.internal.policy.DecorView received View#onDetachedFromWindow() callback) with key fafd8a34-93bc-4bf2-b9b4-4c273382e4ff
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
I/chatty: uid=10134(cgeo.geocaching) RenderThread identical 21 lines
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
W/ActivityThread: handleWindowVisibility: no activity for token android.os.BinderProxy@d4c8808
W/cgeo: [main] No Google API key available.
W/cgeo: [main] No Google API key available.
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
D/LeakCanary: Watching instance of android.widget.PopupWindow$PopupDecorView (android.widget.PopupWindow$PopupDecorView received View#onDetachedFromWindow() callback) with key e854f331-7e03-4c5c-98c0-ac9461281b65
D/EGL_emulation: eglMakeCurrent: 0xe42c2e00: ver 3 0 (tinfo 0xc46f9920)
I/cgeo.geocachin: Explicit concurrent copying GC freed 25094(1211KB) AllocSpace objects, 3(60KB) LOS objects, 49% free, 4738KB/9477KB, paused 1.803ms total 94.119ms
I/cgeo.geocachin: WaitForGcToComplete blocked NativeAlloc on HeapTrim for 88.945ms
D/LeakCanary: Found 2 objects retained, not dumping heap yet (app is visible & < 5 threshold)
I/cgeo.geocachin: Explicit concurrent copying GC freed 12061(672KB) AllocSpace objects, 3(60KB) LOS objects, 49% free, 4534KB/9069KB, paused 784us total 87.290ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 2194(177KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 4466KB/8932KB, paused 388us total 18.306ms
D/eglCodecCommon: allocate: Ask for block of size 0x8d0000
    allocate: ioctl allocate returned offset 0x3fef34000 size 0x8d1000
I/cgeo.geocachin: Explicit concurrent copying GC freed 8009(511KB) AllocSpace objects, 1(20KB) LOS objects, 49% free, 4732KB/9465KB, paused 391us total 46.341ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 1501(175KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 4713KB/9427KB, paused 436us total 56.076ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 1435(162KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 4783KB/9567KB, paused 664us total 29.965ms
D/LeakCanary: Watching instance of com.android.internal.policy.DecorView (com.android.internal.policy.DecorView received View#onDetachedFromWindow() callback) with key 9610b79b-1d15-4d73-92bf-9492de400d3a
I/cgeo.geocachin: Explicit concurrent copying GC freed 1205(189KB) AllocSpace objects, 0(0B) LOS objects, 24% free, 4804KB/6405KB, paused 750us total 23.296ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 6860(389KB) AllocSpace objects, 0(0B) LOS objects, 24% free, 4510KB/6014KB, paused 965us total 32.402ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 471(121KB) AllocSpace objects, 0(0B) LOS objects, 24% free, 4501KB/6002KB, paused 370us total 17.670ms
I/cgeo.geocachin: Explicit concurrent copying GC freed 47(96KB) AllocSpace objects, 0(0B) LOS objects, 24% free, 4501KB/6001KB, paused 806us total 31.391ms
D/LeakCanary: Removing 1 heap dumps
I/cgeo.geocachin: hprof: heap dump "/storage/emulated/0/Download/leakcanary-cgeo.geocaching/2021-02-01_11-36-18_757.hprof" starting...
I/cgeo.geocachin: hprof: heap dump completed (19MB) in 2.098s objects 330748 objects with stack traces 0
I/cgeo.geocachin: Thread[6,tid=14199,WaitingInMainSignalCatcherLoop,Thread*=0xdc220800,peer=0x13100348,"Signal Catcher"]: reacting to signal 3
I/cgeo.geocachin: Wrote stack traces to tombstoned
I/Choreographer: Skipped 1747 frames!  The application may be doing too much work on its main thread.
D/LeakCanary: Analysis in progress, working on: PARSING_HEAP_DUMP
W/cgeo.geocachin: Accessing hidden field Landroid/app/ActivityThread;->mServices:Landroid/util/ArrayMap; (greylist, reflection, allowed)
D/LeakCanary: Watching instance of leakcanary.internal.HeapAnalyzerService (leakcanary.internal.HeapAnalyzerService received Service#onDestroy() callback) with key 25ee0a24-2f82-491c-936e-dc5dc15f1510
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: cgeo.geocaching, PID: 14188
    android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{6e5eb90 u0 cgeo.geocaching/leakcanary.internal.HeapAnalyzerService}
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1945)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

**Comment by fm-sys** [comment]
Edit: All files were moved correctly, so the crash happened after that...

**Comment by fm-sys** [comment]
I checked again: In the old directory, there were remained 5 empty folders inside. The base directory was still set to this old location. After selecting the same new directory again the remaining empty folders were deleted from the old directory and the base directory was changed in c:geo to the new location. 

I hope this information is helpful for debugging...

**Comment by eddiemuc** [comment]
damn. anything "suspicious" about the remaining 5 dirs (name structure, very deep in tree or something like that)?
How many files were there in the tree initially roughly?

progress bar not implemented yet, will follow. But of course the system should not crash.

**Comment by fm-sys** [comment]
> anything "suspicious" about the remaining 5 dirs

They where all directly in the root directory.

Field-notes
themes
backup
gpx
logfiles

-> all existing root folders 

Everything in these folders where moved correctly (also the subdirectories inside backup). The files all together had a size of around 400 MB and I was moving from internal storage to SD. It looks to me (without diving deeper into it) as if moving is done at the UI thread. Could it be that the Android operating system kills the app if it doesn't respond for too long?

**Comment by eddiemuc** [comment]
Yes copying currently runs on UI thread. Working on it.

Would actually be a great relief to me if this is the reason for the crash. Is there such a thing as "app cancelation" if UI thread does not react for longer time?

**Comment by fm-sys** [comment]
As the actual FATAL EXCEPTION is caused by LeakCanary (see above) it would at least sound logical to me... For me it reads like: "couldn't start tread in a reasonable time"

**Comment by bekuno** [comment]
> 
> 
> As the actual FATAL EXCEPTION is caused by LeakCanary (see above) it would at least sound logical to me... For me it reads like: "couldn't start tread in a reasonable time"

Correct, see the row before:
~~~
I/cgeo.geocachin: hprof: heap dump "/storage/emulated/0/Download/leakcanary-cgeo.geocaching/2021-02-01_11-36-18_757.hprof" starting...
I/cgeo.geocachin: hprof: heap dump completed (19MB) in 2.098s objects 330748 objects with stack traces 0
I/cgeo.geocachin: Thread[6,tid=14199,WaitingInMainSignalCatcherLoop,Thread*=0xdc220800,peer=0x13100348,"Signal Catcher"]: reacting to signal 3
I/cgeo.geocachin: Wrote stack traces to tombstoned
I/Choreographer: Skipped 1747 frames!  The application may be doing too much work on its main thread.
D/LeakCanary: Analysis in progress, working on: PARSING_HEAP_DUMP
W/cgeo.geocachin: Accessing hidden field Landroid/app/ActivityThread;->mServices:Landroid/util/ArrayMap; (greylist, reflection, allowed)
D/LeakCanary: Watching instance of leakcanary.internal.HeapAnalyzerService (leakcanary.internal.HeapAnalyzerService received Service#onDestroy() callback) with key 25ee0a24-2f82-491c-936e-dc5dc15f1510
D/AndroidRuntime: Shutting down VM
~~~
especially
**`I/Choreographer: Skipped 1747 frames!  The application may be doing too much work on its main thread.`**
Therefore the system (**Android**) closes the app.

**Comment by bekuno** [comment]
I took a look at the issues list of LeakCanary. There are some issues which deal with the heap dump performance.
But I did not see a planned change.

**Comment by bekuno** [comment]
Maybe we have a chance, see:
https://square.github.io/leakcanary/api/leakcanary-android-process/leakcanary/-leak-canary-process/
Please use in main/build.gradle
`    debugImplementation "com.squareup.leakcanary:leakcanary-android-process:$leakCanaryVersion"`
instead of
`    debugImplementation "com.squareup.leakcanary:leakcanary-android:$leakCanaryVersion"`

**Comment by bekuno** [comment]
The heap dump was initialized because LeakCanary detected potentially leaks:
`D/LeakCanary: Found 2 objects retained, not dumping heap yet (app is visible & < 5 threshold)`

Either there is a real leak or the 5 sec limit was reached.

A initial description for LeakCanary is at https://square.github.io/leakcanary/fundamentals-how-leakcanary-works/

**Comment by eddiemuc** [comment]
Thanks guys, that's a real release! Asynchronous copy change will follow as soon as I am able to code again. Might be Weekend.

**Comment by eddiemuc** [comment]
Should be solved by PR #9955
